derive_builder/src/options/darling_opts.rs

```diff
@@ -0,0 +1,486 @@
+use std::ops;
+use std::vec::IntoIter;
+
+use derive_builder_core::BuildMethod;
+
+use darling::util::{Flag, IdentList};
+use darling::{self, FromMeta};
+use syn::{self, Attribute, Generics, Ident, Path, Visibility};
+
+use derive_builder_core::{
+    Bindings, Builder, BuilderField, BuilderPattern, DeprecationNotes, Initializer, Setter,
+};
+use options::DefaultExpression;
+
+trait FlagVisibility {
+    fn public(&self) -> &Flag;
+    fn private(&self) -> &Flag;
+
+    fn as_expressed_vis(&self) -> Option<Visibility> {
+        match (self.public().is_some(), self.private().is_some()) {
+            (true, true) => panic!("A field cannot be both public and private"),
+            (true, false) => Some(syn::parse_str("pub").unwrap()),
+            (false, true) => Some(Visibility::Inherited),
+            (false, false) => None,
+        }
+    }
+}
+
+/// Options for the `build_fn` property in struct-level builder options.
+#[derive(Debug, Clone, FromMeta)]
+#[darling(default)]
+pub struct BuildFn {
+    skip: bool,
+    name: Ident,
+    validate: Option<Path>,
+}
+
+impl Default for BuildFn {
+    fn default() -> Self {
+        BuildFn {
+            skip: false,
+            name: Ident::from("build"),
+            validate: None,
+        }
+    }
+}
+
+/// Contents of the `field` meta in `builder` attributes.
+#[derive(Debug, Clone, Default, FromMeta)]
+#[darling(default)]
+pub struct FieldMeta {
+    public: Flag,
+    private: Flag,
+}
+
+impl FlagVisibility for FieldMeta {
+    fn public(&self) -> &Flag {
+        &self.public
+    }
+
+    fn private(&self) -> &Flag {
+        &self.private
+    }
+}
+
+#[derive(Debug, Clone, Default, FromMeta)]
+#[darling(default)]
+pub struct StructLevelSetter {
+    prefix: Option<Ident>,
+    into: Option<bool>,
+    skip: Option<bool>,
+}
+
+impl StructLevelSetter {
+    /// Check if setters are explicitly enabled or disabled at
+    /// the struct level.
+    pub fn enabled(&self) -> Option<bool> {
+        self.skip.map(ops::Not::not)
+    }
+}
+
+#[derive(Debug, Clone, Default, FromMeta)]
+#[darling(default)]
+pub struct FieldLevelSetter {
+    prefix: Option<Ident>,
+    name: Option<Ident>,
+    into: Option<bool>,
+    skip: Option<bool>,
+}
+
+impl FieldLevelSetter {
+    /// Get whether or not this field-level setter indicates a setter should
+    /// be emitted. The setter shorthand rules are that the presence of a
+    /// `setter` with _any_ properties set forces the setter to be emitted.
+    pub fn enabled(&self) -> Option<bool> {
+        if self.skip.is_some() {
+            return self.skip.map(ops::Not::not);
+        }
+
+        if self.prefix.is_some() || self.name.is_some() || self.into.is_some() {
+            return Some(true);
+        }
+
+        None
+    }
+}
+
+#[derive(Debug, Clone)]
+enum FieldSetterMeta {
```

Comment:
This needs explanatory comments. Also, it might be worth exploring if this can be generic over `T where T: FromMeta, Shorthand: Into<T>`. I think we only need it once, but it's a useful pattern.